File: CODE/BoxscoresReticulate.R
# Copyright 2025 by Steven J. Keyes. All rights reserved.
# FILE: Example_16_BoxscoresReticulate.R
# DATE 2025-10-17
# DESCRIPTION: 


# R script to call the Python function

# Load the reticulate library
library(reticulate)

# Source the Python script containing the getBoxscores function
# Make sure the file path is correct for your system.
source_python("BoxscoresReticulate.py")

# Define the database and table names.
# Adjust the file path to point to your database.
db <- file.path(Sys.getenv("EXAMPLES"), "Boxscores.db")
table <- 'Boxscores'

# Call the Python function.
# The result will be an R data frame due to reticulate's automatic conversion.
boxscores_df_r <- getBoxscores(db, table)

# Check the class of the returned object to confirm it's an R data frame
print(class(boxscores_df_r))

# Print the first few rows of the data frame
if (!is.null(boxscores_df_r)) {
  print(head(boxscores_df_r))
} else {
  print("Failed to retrieve data. Check the Python function and database path.")
}

print("Done")

File: CODE/BoxscoresReticulate.py
# Copyright 2025 by Steven J. Keyes. All rights reserved.
# FILE: BoxscoresReticulate.py
# DATE 2025-10-16
# DESCRIPTION: 
# This program defines a specialized Python function, getBoxscores, 
# designed to safely extract data from a SQLite database and convert 
# it into a Pandas DataFrame.


import sqlite3
import pandas as pd
import os

def getBoxscores(dbname, table_name):

    # The 'with' statement ensures the connection is properly closed
    try:
        with sqlite3.connect(dbname) as conn:
            # Construct the SQL query to select all data frome table
            query = f"SELECT * FROM {table_name}"
            
            # Use pandas.read_sql_query to execute query and load data
            df = pd.read_sql_query(query, conn)
            
    except sqlite3.Error as e:
        return None
    except pd.io.sql.DatabaseError as e:
        return None
    except FileNotFoundError:
        return None
    except Exception as e:
        return None
    
    return(df)
